insonmnia/miner/builder.go

```diff
@@ -223,15 +229,71 @@ func (b *MinerBuilder) getPublicIPs() (pubIPs []string, err error) {
 	return nil, errors.New("failed to get public IPs")
 }
 
+func (b *MinerBuilder) getHubConnectionInfo() (string, common.Address, error) {
+	var hubEndpoint = b.cfg.HubEndpoint()
+	if strings.Contains(hubEndpoint, "@") {
+		return util.ParseEndpoint(b.cfg.HubEndpoint())
+	} else {
+		resolved, err := b.locatorClient.Resolve(b.ctx, &pb.ResolveRequest{EthAddr: hubEndpoint})
+		if err != nil {
+			return "", common.Address{}, fmt.Errorf(
+				"failed to resolve hub addr from %s: %s", hubEndpoint, err)
+		}
+
+		log.G(b.ctx).Info("resolved hub endpoints", zap.Any("endpoints", resolved.IpAddr))
+
+		for _, addr := range resolved.IpAddr {
+			addr = strings.Replace(addr, "10001", "10002", 1)
+
+			log.G(b.ctx).Debug("trying hub endpoint", zap.Any("endpoint", addr))
+
+			dialer := net.Dialer{DualStack: true, Timeout: time.Second}
+			testCC, err := dialer.DialContext(b.ctx, "tcp", addr)
+			if err != nil {
+				log.G(b.ctx).Debug(
+					"hub endpoint is unreachable", zap.Any("endpoint", addr),
+					zap.Any("error", err))
+			} else {
+				testCC.Close()
+				hubSockaddr, hubEthAddr, err := util.ParseEndpoint(hubEndpoint + "@" + addr)
+				if err == nil {
+					return hubSockaddr, hubEthAddr, nil
+				}
+			}
+		}
+
+		return "", common.Address{}, errors.New("all hub endpoints are unreachable")
+	}
+}
+
 func makeCgroupManager(cfg *ResourcesConfig) (cGroup, cGroupManager, error) {
 	if !platformSupportCGroups || cfg == nil {
 		return newNilCgroupManager()
 	}
 	return newCgroupManager(cfg.Cgroup, cfg.Resources)
 }
 
-func NewMinerBuilder(cfg Config, key *ecdsa.PrivateKey) MinerBuilder {
-	b := MinerBuilder{key: key}
+func NewMinerBuilder(cfg Config, key *ecdsa.PrivateKey) (*MinerBuilder, error) {
+	b := &MinerBuilder{key: key}
 	b.Config(cfg)
-	return b
+
+	if b.ctx == nil {
+		b.ctx = context.Background()
+	}
+
+	_, TLSConf, err := util.NewHitlessCertRotator(b.ctx, b.key)
```

Comment:
do not drop rotator. Its method Stop should be called

Reply:
It's actually dropped in a couple of other places around the project. Should I fix those spots?